### PR TITLE
Implemented `rubi_match()` for rubi integration

### PR DESCRIPTION
**Pull request by arihantparsoya**
This is only a rough implementation of my ideas for Rubi pattern matcher.

`rubi_match()` uses `_partition_args()` function to partition the terms in the expression and match the terms having same head(eg: `Add`, `Mul`, etc).

I compared `rubi_match()` with existing `.match()` function to see how matching time varies with increase in length of expression:

screen shot 2017-05-13 at 15 05 51

The terms used in the above graph are:
```python
tests = [
    [(2*x + 3), (a*x + b)],
    [(2*x + 3)*(4*x + 5), (a*x + b)*(c*x + d)],
    [(2*x + 3)*(4*x + 5)*(6*x + 7), (a*x + b)*(c*x + d)*(e*x + f)],
    [(2*x + 3)*(4*x + 5)*(6*x + 7)*(8*x + 9), (a*x + b)*(c*x + d)*(e*x + f)*(g*x + h)],
    [(2*x + 3)*(4*x + 5)*(6*x + 7)*(8*x + 9)*(10*x + 11), (a*x + b)*(c*x + d)*(e*x + f)*(g*x + h)*(i*x + j)],
]
```
#### Assumptions:
* Patterns are linear(a Wild variable cannot be repeated twice). However, rubi contains patterns that are non-linear. Current implementation can be extended to non-linear patterns also.

#### Notes:
* Basic tests are added in `/pm/test_pm.py`,
* There are many optimization that needs to be done which will make `rubi_match()` faster.

### CODE REVIEW

**Comment by parsoyaarihant** [comment]
@Upabjojr ping.

File: sympy/pm/pm.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+from sympy.utilities.iterables import dict_merge
+from sympy.core.compatibility import _nodes
+
+
+def rubi_match(subject, pattern, var=None):
```
Comment (Upabjojr) [inline sympy/pm/pm.py]: Why `rubi_match`? If we want to create a matching object, it should not be restricted to the RUBI only.

File: bin/generate_module_list.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
     'sympy.functions.combinatorial',
     'sympy.functions.elementary',
     'sympy.functions.elementary.benchmarks',
+    'sympy.pm'
```
Comment (Upabjojr) [inline bin/generate_module_list.py]: `pm` is a too short name. Anyways, I would include this into `sympy.core`.

**Comment by Upabjojr** [comment]
As pointed out by @asmeurer in https://github.com/sympy/sympy/issues/12233#issuecomment-300019128, there is a very interesting project:
https://github.com/HPAC/matchpy

I haven't had time to examine it carefully, but it looks really good. The license is MIT (which is compatible).

File: bin/generate_module_list.py (hunk 2 above)
Comment (debugger22) [inline bin/generate_module_list.py]: I'd also go for something more descriptive.

**Pull request closed without merging**